test/extended/storage/driver_configuration.go

```diff
@@ -0,0 +1,399 @@
+			o.Expect(err).NotTo(o.HaveOccurred(), "failed to get ClusterCSIDriver")
+
+			clusterCSIDriver.Spec.DriverConfig = *originalDriverConfigSpec
+			_, err = oc.AdminOperatorClient().OperatorV1().ClusterCSIDrivers().Update(ctx, clusterCSIDriver, metav1.UpdateOptions{})
+			o.Expect(err).NotTo(o.HaveOccurred(), "failed to update ClusterCSIDriver")
+
+			return err
+		}); err != nil {
+			e2e.Failf("failed to restore original driverConfig of ClusterCSIDriver: %v", err)
+		}
+		e2e.Logf("Successfully restored original driverConfig of ClusterCSIDriver")
+
+	})
+
+	g.Context("snapshot options in clusterCSIDriver should", func() {
+		var tests = []struct {
+			name                       string
+			clusterCSIDriverOptions    *opv1.VSphereCSIDriverConfigSpec
+			cloudConfigOptions         map[string]string
+			successfulSnapshotsCreated int // Number of snapshots that should be created successfully, 0 to skip.
+		}{
+			{
+				name:                       "use default when unset",
+				clusterCSIDriverOptions:    nil,
+				cloudConfigOptions:         map[string]string{},
+				successfulSnapshotsCreated: 3,
+			},
+			{
+				name: "allow setting global snapshot limit",
+				clusterCSIDriverOptions: &opv1.VSphereCSIDriverConfigSpec{
+					GlobalMaxSnapshotsPerBlockVolume: ptr.To(uint32(4)),
+				},
+				cloudConfigOptions: map[string]string{
+					"global-max-snapshots-per-block-volume": "4",
+				},
+				successfulSnapshotsCreated: 4,
+			},
+			{
+				name: "allow setting VSAN limit",
+				clusterCSIDriverOptions: &opv1.VSphereCSIDriverConfigSpec{
+					GranularMaxSnapshotsPerBlockVolumeInVSAN: ptr.To(uint32(4)),
+				},
+				cloudConfigOptions: map[string]string{
+					"granular-max-snapshots-per-block-volume-vsan": "4",
+				},
+				successfulSnapshotsCreated: 0,
+			},
+			{
+				name: "allow setting VVOL limit",
+				clusterCSIDriverOptions: &opv1.VSphereCSIDriverConfigSpec{
+					GranularMaxSnapshotsPerBlockVolumeInVVOL: ptr.To(uint32(4)),
+				},
+				cloudConfigOptions: map[string]string{
+					"granular-max-snapshots-per-block-volume-vvol": "4",
+				},
+				successfulSnapshotsCreated: 0,
+			},
+			{
+				name: "allow all limits to be set at once",
+				clusterCSIDriverOptions: &opv1.VSphereCSIDriverConfigSpec{
+					GlobalMaxSnapshotsPerBlockVolume:         ptr.To(uint32(5)),
+					GranularMaxSnapshotsPerBlockVolumeInVSAN: ptr.To(uint32(10)),
+					GranularMaxSnapshotsPerBlockVolumeInVVOL: ptr.To(uint32(15)),
+				},
+				cloudConfigOptions: map[string]string{
+					"global-max-snapshots-per-block-volume":        "5",
+					"granular-max-snapshots-per-block-volume-vsan": "10",
+					"granular-max-snapshots-per-block-volume-vvol": "15",
+				},
+				successfulSnapshotsCreated: 0,
+			},
+		}
+
+		for _, t := range tests {
+			t := t
+			g.It(fmt.Sprintf("%s", t.name), func() {
+
+				if t.clusterCSIDriverOptions != nil {
```

Comment:
This assumes the ClusterCSIDriver has `Spec.DriverConfig.VSphere == nil`. IMO it would be better to force set `nil` there, overriding any current setting.

Reply:
`t.clusterCSIDriverOptions` is a `VSphereCSIDriverConfigSpec` just for tests to skip setting options if unset/nil (just the first test now). Maybe I misunderstand the suggestion, could you elaborate please?

Reply:
You don't know what's in the original `ClusterCSIDriver` instance present in the cluster. There could be some snapshot setting that the test does not expect and it may allow more (or less) than expected snapshots.

Reply:
I see now, I'll let the test override the setting then. Also adding `RetryOnConflict`  since the update could fail there too. Thanks!